src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -425,42 +422,51 @@ void HRVOAgent::computeNewVelocity()
             firstIntersectingVelocityObstacle(candidate);
-
-        // if this candidate point doesn't intersect anything, use this candidate point
-        if (first_intersecting_velocity_obstacle == -1)
+        bool is_free_of_velocity_obstacle_intersections_and_fast =
+            first_intersecting_velocity_obstacle == -1 &&
+            candidate.velocity.length() >= min_pref_speed;
+        bool
+            is_free_of_velocity_obstacle_intersections_and_slow_but_faster_than_current_selected_speed =
+                first_intersecting_velocity_obstacle == -1 &&
+                candidate.velocity.length() < min_pref_speed &&
+                candidate.velocity.length() > new_velocity_.length();
+        bool
+            is_intersecting_a_velocity_obstacle_further_away_than_the_current_optimal_and_fast =
+                first_intersecting_velocity_obstacle > optimal_furthest_away_obstacle &&
+                candidate.velocity.length() >= min_pref_speed;
+        bool
+            is_intersecting_a_vo_further_away_than_optimal_and_slow_but_faster_than_current_best =
+                first_intersecting_velocity_obstacle == -1 &&
+                candidate.velocity.length() < min_pref_speed &&
+                candidate.velocity.length() >= new_velocity_.length();
+
+        // return as soon as we find a candidate velocity that doesn't intersect anything
+        // and is fast
+        if (is_free_of_velocity_obstacle_intersections_and_fast)
         {
-            new_velocity_ = candidate.position_;
+            new_velocity_ = candidate.velocity;
             return;
         }
 
-        // if this candidate point intersects a velocity obstacle that is further away and
-        // is faster than the minimum preferred speed, then store it as the most optimal
-        // velocity found so far
-        if (first_intersecting_velocity_obstacle > optimal_furthest_away_obstacle &&
-            candidate.position_.length() >= min_pref_speed)
-        {
-            optimal_furthest_away_obstacle = first_intersecting_velocity_obstacle;
-            new_velocity_                  = candidate.position_;
-        }
-        // it's okay to choose a velocity lower than the minimum preferred speed if it's
-        // faster than the currently chosen speed
-        else if (first_intersecting_velocity_obstacle > optimal_furthest_away_obstacle &&
-                 candidate.position_.length() < min_pref_speed &&
-                 candidate.position_.length() >= new_velocity_.length())
+        // if this candidate velocity is flawed, but is better than the one we have so
+        // far, store it until we find something better
+        if (is_free_of_velocity_obstacle_intersections_and_slow_but_faster_than_current_selected_speed ||
+            is_intersecting_a_velocity_obstacle_further_away_than_the_current_optimal_and_fast ||
+            is_intersecting_a_vo_further_away_than_optimal_and_slow_but_faster_than_current_best)
```

Comment:
These have to be longest variable names I've seen yet 😄, but I like that they're descriptive.
nit, `velocity_obstacle` and `vo` seem to be interchanges between the variable names.

Reply:
yeah I rewrite this part as per Jon's suggestion to use functions. if you have any suggestions of making this more readable, go ahead because I couldn't think of anything better